ca/pkcs8/pkcs8.go

```diff
@@ -0,0 +1,247 @@
+	Version             int
+	PrivateKeyAlgorithm []asn1.ObjectIdentifier
+	PrivateKey          []byte
+}
+
+// Encrypted PKCS8
+type pbkdf2Params struct {
+	Salt           []byte
+	IterationCount int
+}
+
+type pbkdf2Algorithms struct {
+	IDPBKDF2     asn1.ObjectIdentifier
+	PBKDF2Params pbkdf2Params
+}
+
+type pbkdf2Encs struct {
+	EncryAlgo asn1.ObjectIdentifier
+	IV        []byte
+}
+
+type pbes2Params struct {
+	KeyDerivationFunc pbkdf2Algorithms
+	EncryptionScheme  pbkdf2Encs
+}
+
+type pbes2Algorithms struct {
+	IDPBES2     asn1.ObjectIdentifier
+	PBES2Params pbes2Params
+}
+
+type encryptedPrivateKeyInfo struct {
+	EncryptionAlgorithm pbes2Algorithms
+	EncryptedData       []byte
+}
+
+// ParsePrivateKeyPEMWithPassword parses PKCS#8 PEM with password to crypto.signer
+func ParsePrivateKeyPEMWithPassword(pemBytes, password []byte) (crypto.Signer, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, errors.New("invalid pem file")
+	}
+
+	var (
+		der []byte
+		err error
+	)
+	der = block.Bytes
+
+	if ok := IsEncryptedPEMBlock(block); ok {
+		der, err = DecryptPEMBlock(block, password)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return derhelpers.ParsePrivateKeyDER(der)
+}
+
+// ParsePrivateKeyPEM parses PKCS#8 PEM to crypto.signer
+func ParsePrivateKeyPEM(pemBytes []byte) (crypto.Signer, error) {
+	block, _ := pem.Decode(pemBytes)
+
+	return derhelpers.ParsePrivateKeyDER(block.Bytes)
+}
+
+// IsEncryptedPEMBlock checks if a PKCS#8 PEM-block is encrypted or not
+func IsEncryptedPEMBlock(block *pem.Block) bool {
+	der := block.Bytes
+
+	var privKey encryptedPrivateKeyInfo
+	if _, err := asn1.Unmarshal(der, &privKey); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// DecryptPEMBlock requires PKCS#8 PEM Block and password to decrypt and return unencrypted der []byte
+func DecryptPEMBlock(block *pem.Block, password []byte) ([]byte, error) {
+	der := block.Bytes
+
+	var privKey encryptedPrivateKeyInfo
+	if _, err := asn1.Unmarshal(der, &privKey); err != nil {
+		return nil, errors.New("pkcs8: only PKCS #5 v2.0 supported")
+	}
+
+	if !privKey.EncryptionAlgorithm.IDPBES2.Equal(oidPBES2) {
+		return nil, errors.New("pkcs8: only PBES2 supported")
+	}
+
+	if !privKey.EncryptionAlgorithm.PBES2Params.KeyDerivationFunc.IDPBKDF2.Equal(oidPKCS5PBKDF2) {
+		return nil, errors.New("pkcs8: only PBKDF2 supported")
+	}
+
+	encParam := privKey.EncryptionAlgorithm.PBES2Params.EncryptionScheme
+	kdfParam := privKey.EncryptionAlgorithm.PBES2Params.KeyDerivationFunc.PBKDF2Params
+
+	switch {
+	case encParam.EncryAlgo.Equal(oidAES256CBC):
+		iv := encParam.IV
+		salt := kdfParam.Salt
+		iter := kdfParam.IterationCount
+
+		encryptedKey := privKey.EncryptedData
+		symkey := pbkdf2.Key(password, salt, iter, 32, sha1.New)
+		block, err := aes.NewCipher(symkey)
+		if err != nil {
+			return nil, err
+		}
+		mode := cipher.NewCBCDecrypter(block, iv)
+		mode.CryptBlocks(encryptedKey, encryptedKey)
+
+		// Remove padding
```

Comment:
Just for my own understanding, when comparing this to https://github.com/docker/notary/pull/1130/files#diff-14d2480b5aeb5325dbdc2c70ae8fd052R215, and to https://github.com/youmark/pkcs8/blob/master/pkcs8.go#L143, why does this implementation need to remove padding?

I assume for `notary`, perhaps the subsequence `asn1.Unmarshal` after the `mode.CryptBlocks` call ignores the extra padding.  It seems like [`x509.ParsePKCS8PrivateKey`](https://golang.org/src/crypto/x509/pkcs8.go?s=580:646#L16) also does an unmarshal immediately, though, and [`derhelpers.ParsePrivateKeyDER`](https://github.com/cloudflare/cfssl/blob/master/helpers/derhelpers/derhelpers.go#L16) calls `x509.ParsePKCS8PrivateKey` immediately.